### PR TITLE
Wrapped classes in discrete

### PR DESCRIPTION
**Pull request by dellaert**
This adds python/MATLAB wrappers for the classes in discrete. @ProfFan helped with some of the wrapper magic.

Development was driven by
- test_DiscreteFactorGraph.py
- test_DiscreteBayesNet.py

To make this possible, I had to do some constructor forwarding etc, but the main new functionality is the ability to construct DiscreteConditionals with a simpler constructor that could be wrapped. You can now do
```
  DiscreteConditional pXE(X, {E}, "95/5 2/98");
```
which complements the *magical* sugar that was there before:
```
  DiscreteConditional pXE(X | E = "95/5 2/98");
```

I also added a variant of DiscreteConditional::choose that returns a DecisionTreeFactor.

It works. Here is code that builds the Asia network:

```python
asia = DiscreteBayesNet()
asia.add(Asia, P([]), "99/1")
asia.add(Smoking, P([]), "50/50")

asia.add(Tuberculosis, P([Asia]), "99/1 95/5")
asia.add(LungCancer, P([Smoking]), "99/1 90/10")
asia.add(Bronchitis, P([Smoking]), "70/30 40/60")

asia.add(Either, P([Tuberculosis, LungCancer]), "F T T T")

asia.add(XRay, P([Either]), "95/5 2/98")
asia.add(Dyspnea, P([Either, Bronchitis]), "9/1 2/8 3/7 1/9")
```

`print(asia)` yields:

```
DiscreteBayesNet

size: 8
factor 0:  P( 0 )
  Cardinalities: {0:2, }
  Choice(0) 
  0 Leaf 0.99
  1 Leaf 0.01

factor 1:  P( 4 )
  Cardinalities: {4:2, }
  Leaf 0.5

factor 2:  P( 3 | 0 )
  Cardinalities: {0:2, 3:2, }
  Choice(3) 
  0 Choice(0) 
  0 0 Leaf 0.99
  0 1 Leaf 0.95
  1 Choice(0) 
  1 0 Leaf 0.01
  1 1 Leaf 0.05

factor 3:  P( 6 | 4 )
  Cardinalities: {4:2, 6:2, }
  Choice(6) 
  0 Choice(4) 
  0 0 Leaf 0.99
  0 1 Leaf 0.9
  1 Choice(4) 
  1 0 Leaf 0.01
  1 1 Leaf 0.1

factor 4:  P( 7 | 4 )
  Cardinalities: {4:2, 7:2, }
  Choice(7) 
  0 Choice(4) 
  0 0 Leaf 0.7
  0 1 Leaf 0.4
  1 Choice(4) 
  1 0 Leaf 0.3
  1 1 Leaf 0.6

factor 5:  P( 5 | 3 6 )
  Cardinalities: {3:2, 5:2, 6:2, }
  Choice(6) 
  0 Choice(5) 
  0 0 Choice(3) 
  0 0 0 Leaf 1
  0 0 1 Leaf 0
  0 1 Choice(3) 
  0 1 0 Leaf 0
  0 1 1 Leaf 1
  1 Choice(5) 
  1 0 Leaf 0
  1 1 Leaf 1

factor 6:  P( 2 | 5 )
  Cardinalities: {2:2, 5:2, }
  Choice(5) 
  0 Choice(2) 
  0 0 Leaf 0.95
  0 1 Leaf 0.05
  1 Choice(2) 
  1 0 Leaf 0.02
  1 1 Leaf 0.98

factor 7:  P( 1 | 5 7 )
  Cardinalities: {1:2, 5:2, 7:2, }
  Choice(7) 
  0 Choice(5) 
  0 0 Choice(1) 
  0 0 0 Leaf 0.9
  0 0 1 Leaf 0.1
  0 1 Choice(1) 
  0 1 0 Leaf 0.3
  0 1 1 Leaf 0.7
  1 Choice(5) 
  1 0 Choice(1) 
  1 0 0 Leaf 0.2
  1 0 1 Leaf 0.8
  1 1 Choice(1) 
  1 1 0 Leaf 0.1
  1 1 1 Leaf 0.9
```

### CODE REVIEW

File: gtsam/discrete/DiscreteConditional.h [hunk 1]
```diff
@@ -57,6 +57,29 @@ class GTSAM_EXPORT DiscreteConditional: public DecisionTreeFactor,
   /** Construct from signature */
   DiscreteConditional(const Signature& signature);
 
+  /**
+  * Construct from key, parents, and a Table specifying the CPT.
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteConditional.h:61]: Same deal here about the CPT. We know what the context is, but so should other people. :)

File: gtsam/discrete/discrete.i [hunk 2]
```diff
@@ -0,0 +1,131 @@
+//*************************************************************************
+// basis
```
Comment (varunagrawal) [inline gtsam/discrete/discrete.i:2]: This should be `discrete`.

File: gtsam/discrete/tests/testDiscreteConditional.cpp [hunk 3]
```diff
@@ -61,11 +61,10 @@ TEST(DiscreteConditional, constructors_alt_interface) {
   r2 += 2.0, 3.0;
   r3 += 1.0, 4.0;
   table += r1, r2, r3;
-  auto actual1 = boost::make_shared<DiscreteConditional>(X | Y = table);
-  EXPECT(actual1);
+  DiscreteConditional actual1(X, {Y}, table);
```
Comment (varunagrawal) [inline gtsam/discrete/tests/testDiscreteConditional.cpp:64]: Amazing :fire:

File: python/gtsam/specializations/discrete.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ * NOTES
+ * =================
+ *
+ * `py::bind_vector` and similar machinery gives the std container a Python-like
+ * interface, but without the `<pybind11/stl.h>` copying mechanism. Combined
+ * with `PYBIND11_MAKE_OPAQUE` this allows the types to be modified with Python,
+ * and saves one copy operation.
+ */
+
+// Seems this is not a good idea with inherited stl
```
Comment (varunagrawal) [inline python/gtsam/specializations/discrete.h:14]: Yeah we have the same issue with `IndexPair` which inherits from `std::pair`. I haven't tried looking too deeply into it.

**Review by varunagrawal — APPROVED** [review]
LGTM module some comments. I'll pull this branch and investigate what the CI failure is.

File: gtsam/discrete/DiscreteFactorGraph.h [hunk 5]
```diff
@@ -130,8 +130,14 @@ public EliminateableFactorGraph<DiscreteFactorGraph> {
...
-  /** Evaluates the factor graph given values, returns the joint probability of the factor graph given specific instantiation of values*/
-  double operator()(const DiscreteValues & values) const;
+  /** 
+   * Evaluates the factor graph given values, returns the joint probability of
+   * the factor graph given specific instantiation of values
+   */
+  double operator()(const DiscreteValues& values) const;
+
+  /// Synonym for operator(), mostly for wrapper
+  double evaluate(const DiscreteValues& values) const { return operator()(values); }
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/discrete/DiscreteFactorGraph.h:140]: Example?

File: gtsam/discrete/Signature.h [hunk 6]
```diff
@@ -72,45 +72,66 @@ namespace gtsam {
...
-    /** Add the CPT spec - Fails in boost 1.40 */
-    Signature& operator=(const std::string& spec);
-
-    /** Add the CPT spec directly as a table */
-    Signature& operator=(const Table& table);
-
-    /** provide streaming */
-    GTSAM_EXPORT friend std::ostream& operator <<(std::ostream &os, const Signature &s);
+   /**
+    * Construct from key, parents, and a Table specifying the CPT.
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/discrete/Signature.h:76]: Added comment.

File: gtsam/discrete/DiscreteFactorGraph.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/discrete/DiscreteFactorGraph.h:140]: The syntax would be `double operator()(const gtsam::DiscreteValues& values) const;`

I used this in the `DecisionTreeFactor` and it works as expected. :slightly_smiling_face: 

```python
In [1]: f = gtsam.DecisionTreeFactor()

In [2]: f
Out[2]: 
DecisionTreeFactor
Potentials:
  Cardinalities: {}
  Leaf 1

In [3]: f()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-c43e34e6d405> in <module>
----> 1 f()

TypeError: __call__(): incompatible function arguments. The following argument types are supported:
    1. (self: gtsam.gtsam.DiscreteFactor, arg0: gtsam::Assignment<unsigned long>) -> float

Invoked with: DecisionTreeFactor
Potentials:
  Cardinalities: {}
  Leaf 1


In [4]: values = gtsam.DiscreteValues()

In [5]: f(values)
Out[5]: 1.0

In [6]: 
```

File: gtsam/discrete/DiscreteConditional.h (hunk 1 above)
Comment (dellaert) [reply gtsam/discrete/DiscreteConditional.h:61]: Good point

File: gtsam/discrete/discrete.i [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  DiscreteConditional(size_t nFrontals, const gtsam::DecisionTreeFactor& f);
+  DiscreteConditional(const gtsam::DiscreteKey& key,
+                      const gtsam::DiscreteKeys& parents, string spec);
+  DiscreteConditional(const gtsam::DecisionTreeFactor& joint,
+                      const gtsam::DecisionTreeFactor& marginal);
+  DiscreteConditional(const gtsam::DecisionTreeFactor& joint,
+                      const gtsam::DecisionTreeFactor& marginal,
+                      const gtsam::Ordering& orderedKeys);
+  size_t size() const; // TODO(dellaert): why do I have to repeat???
+  double evaluate(const gtsam::DiscreteValues& values) const; // TODO(dellaert): why do I have to repeat???
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/discrete/discrete.i:56]: NOOOOO! That worked before. So, the semantics changed... The wrapper should generate all wrapped functions in base classes I think.

File: gtsam/discrete/discrete.i (hunk 2 above)
Comment (dellaert) [reply gtsam/discrete/discrete.i:2]: Done

File: python/gtsam/specializations/discrete.h (hunk 4 above)
Comment (dellaert) [reply python/gtsam/specializations/discrete.h:14]: Would be amazing if this works.

**Comment by varunagrawal** [comment]
@dellaert I fixed the CI issue in #968

**Comment by varunagrawal** [comment]
I also added the GTSAM_EXPORT declarations in #968 so that Windows is happy.

File: gtsam/discrete/discrete.i (hunk 7 above)
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtsam/discrete/discrete.i]: Gotcha. Okay then, we'll log this as a feature request.

File: python/gtsam/specializations/discrete.h (hunk 4 above)
Comment (varunagrawal) [reply python/gtsam/specializations/discrete.h:14]: I found this as an undocumented feature: https://github.com/pybind/pybind11/issues/620

**Pull request merged**